dbms/src/Flash/Mpp/HashBaseWriterHelper.cpp

```diff
@@ -41,44 +41,94 @@ std::vector<MutableColumns> createDestColumns(const Block & sample_block, size_t
     return dest_tbl_cols;
 }
 
+void fillSelector(size_t rows,
+                  const WeakHash32 & hash,
+                  uint32_t part_num,
+                  IColumn::Selector & selector)
+{
+    // fill selector array with most significant bits of hash values
+    const auto & hash_data = hash.getData();
+    selector.resize(rows);
+    for (size_t i = 0; i < rows; ++i)
+    {
+        /// Row from interval [(2^32 / part_num) * i, (2^32 / part_num) * (i + 1)) goes to partition with number i.
+        selector[i] = hash_data[i]; /// [0, 2^32)
+        selector[i] *= part_num; /// [0, part_num * 2^32), selector stores 64 bit values.
+        selector[i] >>= 32u; /// [0, part_num)
+    }
+}
+
+/// For FineGrainedShuffle, the selector algorithm should satisfy the requirement:
+//  the FineGrainedShuffleStreamIndex can be calculated using hash_data and fine_grained_shuffle_stream_count values, without the presence of part_num.
+void fillSelectorForFineGrainedShuffle(size_t rows,
+                                       const WeakHash32 & hash,
+                                       uint32_t part_num,
+                                       uint32_t fine_grained_shuffle_stream_count,
+                                       IColumn::Selector & selector)
+{
+    // fill selector array with most significant bits of hash values
+    const auto & hash_data = hash.getData();
+    selector.resize(rows);
+    for (size_t i = 0; i < rows; ++i)
+    {
+        /// Row from interval [(2^32 / part_num) * i, (2^32 / part_num) * (i + 1)) goes to partition with number i.
+        selector[i] = hash_data[i]; /// [0, 2^32)
+        selector[i] *= part_num; /// [0, part_num * 2^32), selector stores 64 bit values.
+        selector[i] >>= 32u; /// [0, part_num)
+        selector[i] = selector[i] * fine_grained_shuffle_stream_count + hash_data[i] % fine_grained_shuffle_stream_count; /// map to [0, part_num * fine_grained_shuffle_stream_count]
```

Comment:
Does this have a negative impact on performance?

Reply:
> Is it `[0, part_num * fine_grained_shuffle_stream_count]` or `[0, part_num * fine_grained_shuffle_stream_count)`?

Yes, you're right, will change it.

Reply:
> Does this have a negative impact on performance?

Yes, it will cost a little more cpu resources, I expect it would not be noticable in total execution time level.

Reply:
> > Is it `[0, part_num * fine_grained_shuffle_stream_count]` or `[0, part_num * fine_grained_shuffle_stream_count)`?
> 
> Yes, you're right, will change it.

Done.